rrdd/rrdd_server.ml

```diff
@@ -415,36 +416,44 @@ module Plugin = struct
 		 * conditions and data corruption. *)
 		let registered_m = Mutex.create ()
 
-		(* Helper function to find plugin info from a uid. *)
-		let find ~(uid: P.uid) : plugin =
-			try
-				Mutex.execute registered_m (fun () -> Hashtbl.find registered uid)
-			with Not_found ->
-				let msg = Printf.sprintf "failed to find plugin %s at %s"
-					(P.string_of_uid uid) __LOC__ in
-				error "internal error: %s" msg;
-				failwith msg
-
-		let mute uid plugin =
-			Mutex.execute registered_m (fun () -> plugin.muted <- true);
-			warn "muting plugin %s" (P.string_of_uid uid)
-
-		let unmute uid plugin =
-			if plugin.muted then begin
-				Mutex.execute registered_m (fun () -> plugin.muted <- false);
-				warn "unmuting plugin %s" (P.string_of_uid uid)
-			end else
-				()
-
-		let get_payload ~(uid: P.uid) : Rrd_protocol.payload =
-			let plugin  = find uid in
+		(* we hit an error - increase skip count exponentially until max *)
+		let incr_skip_count uid plugin =
+			let skip_max  = 256 in (* about 21.3 min on 5sec cycle *)
+			let min x y   = if x < y then x else y in
+			Mutex.execute registered_m (fun () ->
+				let skips     = min (plugin.skip_init * 2) skip_max in
+				warn "setting skip-cycles-after-error for plugin %s to %d"
+					(P.string_of_uid uid) skips;
+				plugin.skip_init <- skips;
+				plugin.skip      <- skips)
+
+		(* success - set skip to 0, reset initial value *)
+		let reset_skip_count uid plugin =
+			if plugin.skip_init > 1 then begin
+				warn "re-setting skip-cycles-after-error for plugin %s to 1"
+					(P.string_of_uid uid);
+				Mutex.execute registered_m  (fun () ->
+					plugin.skip_init <- 1;
+					plugin.skip      <- 0)
+			end
+
+		(* we are skipping a reading *)
+		let decr_skip_count (uid, plugin) =
+			if plugin.skip > 0 then
+				Mutex.execute registered_m  (fun () ->
+					plugin.skip <- plugin.skip - 1)
+
+		(* true, iff the plugin skips the next reading *)
+		let skip (uid, plugin) = plugin.skip > 0
```

Comment:
Ah right, I missed it. I think you could define it earlier and use it in `decr_skip_count` (line 442) as well